src/core/core.element.js

```diff
@@ -1,7 +1,77 @@
 import {inherits} from '../helpers/helpers.core';
 import {isNumber} from '../helpers/helpers.math';
+import {resolve} from '../helpers/helpers.options';
+import color from '@kurkle/color';
 
-class Element {
+/**
+ * @typedef { import("./core.controller").default } Chart
+ */
+
+const ucFirst = (str) => str.charAt(0).toUpperCase() + str.slice(1);
+
+export const getHoverColor = function(colorValue) {
+	return colorValue && ((colorValue instanceof CanvasPattern || colorValue instanceof CanvasGradient)
+		? colorValue
+		: color(colorValue).saturate(0.5).darken(0.1).hexString());
+};
+
+const notIndexable = ['borderDash', 'fill'];
+const indexable = (key, index) => notIndexable.indexOf(key) !== -1 ? undefined : index;
+
+export default class Element {
+
+	static _type = 'element';
+
+	static extend = inherits;
+
+	/**
+	 * Default options. Everything is looked up with `hover` prefix, so
+	 * only need to define those in case the default value is different.
+	 */
+	static _defaults = {};
+
+	/**
+	 * @param {typeof Element} elementType
+	 * @param {object[]} inputs
+	 * @param {{ chart?: Chart; dataIndex?: number; dataset?: any; datasetIndex?: number; active?: boolean; index?: number; }} context
+	 * @param {string} mode
+	 */
+	static resolveOptions(elementType, inputs, context, mode) {
```

Comment:
I kind of have this mental model that the controller does most of the computation and the element is more of a data model, so I think it'd be nice to keep this in datasetController since that's the only place it's called

Reply:
I agree here. To me, keeping the models simpler makes it a lot easier to reason about for me

Reply:
Ok, how do about moving this to helpers?
Just wanting to make the datasetController a bit less bulky.

Reply:
Functions to do this works for me, also a bit easier to test in isolation